extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ShutdownBuildSteps.java

```diff
@@ -0,0 +1,130 @@
+            ClassInfo beanClass = bean.getTarget().get().asClass();
+            List<MethodInfo> shutdownMethods = new ArrayList<>();
+            // Collect all non-static no-args methods annotated with @Shutdown
+            for (MethodInfo method : beanClass.methods()) {
+                if (annotationStore.hasAnnotation(method, SHUTDOWN_NAME)) {
+                    if (!method.isSynthetic()
+                            && !Modifier.isStatic(method.flags())
+                            && method.parametersCount() == 0) {
+                        shutdownMethods.add(method);
+                    } else {
+                        throw new IllegalStateException(
+                                "Invalid @Shutdown method declared on " + method.declaringClass().name() + ": " + method);
+                    }
+                }
+            }
+            if (!shutdownMethods.isEmpty()) {
+                for (MethodInfo method : shutdownMethods) {
+                    AnnotationValue priority = annotationStore.getAnnotation(method, SHUTDOWN_NAME).value();
+                    registerShutdownObserver(observerRegistration, bean,
+                            method.declaringClass().name() + "#" + method.toString(),
+                            priority != null ? priority.asInt() : ObserverMethod.DEFAULT_PRIORITY, method);
+                }
+            }
+        }
+    }
+
+    private void registerShutdownObserver(ObserverRegistrationPhaseBuildItem observerRegistration, BeanInfo bean, String id,
+            int priority, MethodInfo shutdownMethod) {
+        ObserverConfigurator configurator = observerRegistration.getContext().configure()
+                .beanClass(bean.getBeanClass())
+                .observedType(ShutdownEvent.class);
+        configurator.id(id);
+        configurator.priority(priority);
+        configurator.notify(mc -> {
+            // InjectableBean<Foo> bean = Arc.container().bean("bflmpsvz");
+            ResultHandle containerHandle = mc.invokeStaticMethod(StartupBuildSteps.ARC_CONTAINER);
+            ResultHandle beanHandle = mc.invokeInterfaceMethod(StartupBuildSteps.ARC_CONTAINER_BEAN, containerHandle,
+                    mc.load(bean.getIdentifier()));
+            if (BuiltinScope.DEPENDENT.is(bean.getScope())) {
+                ResultHandle creationalContext = mc.newInstance(
+                        MethodDescriptor.ofConstructor(CreationalContextImpl.class, Contextual.class),
+                        beanHandle);
+                // Create a dependent instance
+                ResultHandle instance = mc.invokeInterfaceMethod(StartupBuildSteps.CONTEXTUAL_CREATE, beanHandle,
+                        creationalContext);
+                mc.invokeVirtualMethod(MethodDescriptor.of(shutdownMethod), instance);
+                // Destroy the instance immediately
+                mc.invokeInterfaceMethod(StartupBuildSteps.CONTEXTUAL_DESTROY, beanHandle, instance, creationalContext);
```

Comment:
This won't work if the shutdown method throws an exception, but that's probably not a big deal.

Reply:
Not sure what we do for startup methods, probably the same? :-)